test/platform.test.ts

```diff
@@ -3,119 +3,596 @@
  *--------------------------------------------------------*/
 
 import * as assert from "assert";
+import * as child_process from "child_process";
+import mockFS = require("mock-fs");
+import FileSystem = require("mock-fs/lib/filesystem");
+import * as path from "path";
+import * as sinon from "sinon";
 import * as platform from "../src/platform";
 
-function checkDefaultPowerShellPath(platformDetails, expectedPath) {
-    test("returns correct default path", () => {
-        assert.equal(
-            platform.getDefaultPowerShellPath(platformDetails),
-            expectedPath);
-    });
+/**
+ * Describes a platform on which the PowerShell extension should work,
+ * including the test conditions (filesystem, environment variables).
+ */
+interface ITestPlatform {
+    name: string;
+    platformDetails: platform.IPlatformDetails;
+    filesystem: FileSystem.DirectoryItems;
+    environmentVars?: Record<string, string>;
 }
 
-function checkAvailableWindowsPowerShellPaths(
-    platformDetails: platform.IPlatformDetails,
-    expectedPaths: platform.IPowerShellExeDetails[]) {
-    test("correctly enumerates available Windows PowerShell paths", () => {
-
-        // The system may return PowerShell Core paths so only
-        // enumerate the first list items.
-        const enumeratedPaths = platform.getAvailablePowerShellExes(platformDetails, undefined);
-        for (let i; i < expectedPaths.length; i++) {
-            assert.equal(enumeratedPaths[i], expectedPaths[i]);
-        }
-    });
+/**
+ * A platform where the extension should find a PowerShell,
+ * including the sequence of PowerShell installations that should be found.
+ * The expected default PowerShell is the first installation.
+ */
+interface ITestPlatformSuccessCase extends ITestPlatform {
+    expectedPowerShellSequence: platform.IPowerShellExeDetails[];
 }
 
-function checkFixedWindowsPowerShellpath(platformDetails, inputPath, expectedPath) {
-    test("fixes incorrect Windows PowerShell Sys* path", () => {
-        assert.equal(
-            platform.fixWindowsPowerShellPath(inputPath, platformDetails),
-            expectedPath);
-    });
-}
+// Platform configurations where we expect to find a set of PowerShells
+let successTestCases: ITestPlatformSuccessCase[];
 
-suite("Platform module", () => {
-    if (process.platform === "win32") {
-        suite("64-bit Windows, 64-bit VS Code", () => {
-            const platformDetails: platform.IPlatformDetails = {
+const pwshMsixDir = "C:\\Program Files\\WindowsApps\\Microsoft.PowerShell_7.0.0.4_neutral__8wekyb3d8bbwe";
+const pwshMsixPath = path.join(pwshMsixDir, "pwsh.exe");
```

Comment:
why the join here?

Reply:
TSLint doesn't like line length, and pwshMsixDir gets used in some test cases